### PR TITLE
perf(core): Add session start profiler

### PR DESCRIPTION
**Pull request by doudouOUC**
## What this PR does

This PR adds an opt-in, core-internal profiler for session startup. When `QWEN_CODE_PROFILE_SESSION_START=1` is set, `GeminiClient.startChat()` records bounded JSONL stage timings under the runtime output directory so session initialization can be broken down without enabling debug logs or changing normal behavior.

The recorded data is intentionally limited to static stage names, source, success state, total and per-stage durations, and small aggregate counts. It avoids prompts, paths, session IDs, model responses, hook output, and tool names.

## Why it's needed

Issue #6312 needs the next optimization to be chosen from measured evidence. Recent measurement showed the remaining light-config cost sits inside `GeminiClient.initialize()`, but it did not identify which `startChat()` stage dominates. This PR adds the measurement hook needed to decide whether to optimize system instruction construction, initial history assembly, tool syncing, hooks, or another stage.

### How to verify

Run the focused core tests and confirm the profiler helper records only when enabled, preserves errors, swallows output-write failures, and that `startChat()` finalizes both success and failure profiles with bounded counts.

Optionally run a local initialization with `QWEN_CODE_PROFILE_SESSION_START=1` and an isolated `QWEN_RUNTIME_DIR`; confirm a `session-start-perf/session-start-YYYY-MM-DD.jsonl` file is produced and contains static stage names plus aggregate counts only.

### Evidence (Before & After)

N/A for UI. Local verification completed with `cd packages/core && npx vitest run src/core/session-start-profiler.test.ts src/core/client.test.ts` passing 236 tests, `npm run build && npm run typecheck` passing, and `npm run lint` passing. A local 12-iteration profiler run was posted to #6312 and showed `system_instruction` as the dominant light-fixture stage.

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  | ✅ tested |
| 🪟 Windows | ⚠️ not tested |
|  🐧 Linux  | ⚠️ not tested |

### Environment (optional)

Node v22.22.3, npm 10.9.8. Local verification used built dist output and isolated `QWEN_HOME` / `QWEN_RUNTIME_DIR` for the profiler smoke measurement.

## Risk & Scope

- Main risk or tradeoff: The instrumentation adds wrapper calls inside `startChat()`, but the helper is disabled unless `QWEN_CODE_PROFILE_SESSION_START=1` and performs no file writes or high-resolution clock reads when disabled.
- Not validated / out of scope: This PR does not optimize `GeminiClient.initialize()` or implement Part B extension caching, Part C skill body lazy-loading, daemon protocol changes, or an end-to-end `/session` latency benchmark.
- Breaking changes / migration notes: None; no public API, CLI flag, SDK schema, config schema, protocol, or telemetry schema changes are introduced.

## Linked Issues

Refs #6312

<details>
<summary>中文说明</summary>

## What this PR does

这个 PR 增加了一个 core 内部的可选 session startup profiler。设置 `QWEN_CODE_PROFILE_SESSION_START=1` 后，`GeminiClient.startChat()` 会在 runtime output 目录下记录有边界的 JSONL 阶段耗时，这样可以拆解 session 初始化路径，而不需要启用 debug log，也不改变正常行为。

记录的数据有意限制在静态阶段名、来源、成功状态、总耗时/阶段耗时以及少量聚合计数。输出不会包含 prompt、路径、session ID、模型响应、hook 输出或工具名。

## Why it's needed

#6312 的下一步优化需要基于实测证据选择。最近的测量显示轻配置下剩余成本位于 `GeminiClient.initialize()` 内，但还没有定位到 `startChat()` 的具体主耗时阶段。这个 PR 增加所需的测量 hook，用于判断后续应优化 system instruction 构造、initial history assembly、tool syncing、hooks，还是其他阶段。

### How to verify

运行聚焦 core 测试，确认 profiler helper 只在启用时记录、保留原错误、吞掉输出写入失败，并确认 `startChat()` 在成功和失败路径都会用有边界的计数完成 profile。

也可以设置 `QWEN_CODE_PROFILE_SESSION_START=1` 并使用隔离的 `QWEN_RUNTIME_DIR` 跑一次本地初始化；预期会生成 `session-start-perf/session-start-YYYY-MM-DD.jsonl`，且内容只包含静态阶段名和聚合计数。

### Evidence (Before & After)

UI 变化不适用。本地验证已完成：`cd packages/core && npx vitest run src/core/session-start-profiler.test.ts src/core/client.test.ts` 通过 236 个测试，`npm run build && npm run typecheck` 通过，`npm run lint` 通过。一次本地 12 轮 profiler 运行结果已评论到 #6312，显示轻配置 fixture 中 `system_instruction` 是主要阶段。

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  | ✅ tested |
| 🪟 Windows | ⚠️ not tested |
|  🐧 Linux  | ⚠️ not tested |

### Environment (optional)

Node v22.22.3，npm 10.9.8。本地验证使用 built dist 输出，并为 profiler smoke measurement 使用隔离的 `QWEN_HOME` / `QWEN_RUNTIME_DIR`。

## Risk & Scope

- Main risk or tradeoff: 这个 instrumentation 在 `startChat()` 内增加了 wrapper 调用，但 helper 只有在 `QWEN_CODE_PROFILE_SESSION_START=1` 时才启用；禁用时不会写文件，也不会读取高精度时钟。
- Not validated / out of scope: 这个 PR 不优化 `GeminiClient.initialize()`，不实现 Part B extension caching、Part C skill body lazy-loading、daemon protocol changes，也不是端到端 `/session` 延迟基准。
- Breaking changes / migration notes: 无；没有引入 public API、CLI flag、SDK schema、config schema、protocol 或 telemetry schema 变更。

## Linked Issues

Refs #6312

</details>

### CODE REVIEW

File: packages/core/src/core/session-start-profiler.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  finish(attrs: SessionStartProfileFinishAttrs): void {
+    if (this.finished) {
+      return;
+    }
+    this.finished = true;
+
+    const record: SessionStartProfileRecord = {
+      timestamp: this.getTimestamp().toISOString(),
```
Comment (wenshao) [inline packages/core/src/core/session-start-profiler.ts]: **[Critical]** `finish()` constructs the record object — calling `this.getTimestamp()` (line 131) and `this.now()` (line 134) — **outside** the try/catch that only wraps `this.writeRecord()` (line 153). If either call throws (e.g., a custom `now`/`getTimestamp` override in tests or a future runtime condition), the error escapes `finish()` unhandled.

In the `startChat()` error path (`client.ts` catch block), `profiler.finish({ ok: false, ... })` runs before `reportError()` and the re-throw. A throwing `finish()` would **mask the original session initialization error** — the caller sees a profiler crash instead of `"Failed to initialize chat: ..."`. This violates the stated design principle: "All profiler writes are best-effort. File-system failures are swallowed so profiling cannot break or slow a session through error handling."

```suggestion
  finish(attrs: SessionStartProfileFinishAttrs): void {
    if (this.finished) {
      return;
    }
    this.finished = true;

    try {
      const record: SessionStartProfileRecord = {
        timestamp: this.getTimestamp().toISOString(),
        source: this.source,
        ok: attrs.ok,
        totalMs: roundMs(this.now() - this.startMs),
        stages: { ...this.stages },
        ...(attrs.extraHistoryLength !== undefined
          ? { extraHistoryLength: attrs.extraHistoryLength }
          : {}),
        ...(attrs.historyLength !== undefined
          ? { historyLength: attrs.historyLength }
          : {}),
        ...(attrs.snapshotEntryCount !== undefined
          ? { snapshotEntryCount: attrs.snapshotEntryCount }
          : {}),
        ...(attrs.deferredReminderCount !== undefined
          ? { deferredReminderCount: attrs.deferredReminderCount }
          : {}),
        ...(this.failedStage ? { failedStage: this.failedStage } : {}),
      };

      this.writeRecord(record);
    } catch {
      // Profiling must never affect session creation.
    }
  }
```

_— qwen3.7-max via Qwen Code /review_

**Review by wenshao — CHANGES_REQUESTED** [review]

**Comment by wenshao** [comment]
### Suggestions — commit `af1016b`

| File | Issue | Suggested fix |
| --- | --- | --- |
| `session-start-profiler.ts:80-93` | `assertSafeExistingProfileFile` never tests the append happy path (second session same day) — only ENOENT and symlink branches are covered | Add a test creating a real JSONL file, then running a second profiler that appends to the same file |
| `client.ts:1357` | `finishProfile(false)` with non-zero partial counts untested — all failure tests fail before `deferredReminderCount`/`snapshotEntries` are populated | Add a test where `setTools` rejects, asserting non-zero partial metadata |
| `session-start-profiler.ts:80-93` | Inline error-code narrowing (`typeof error === 'object' && 'code' in error`) should use `isNodeError()` from `../utils/errors.js` for consistency | Import and use `isNodeError()` for both `assertSafeExistingProfileFile` and `finish()` catch blocks |
| `session-start-profiler.ts:213-218` | `finish()` catch logs at `debug` level without target path or directory context — operators diagnosing profiling failures lack key info | Log at `warn` level and include `source` and output directory in the log payload |
| `session-start-profiler.ts:237-248` | No activation log when profiler is enabled — misspelled env var is indistinguishable from write failure | Emit `debugLogger.info('session-start-profiler enabled', { source })` on creation |
| `session-start-profiler.ts:197` | `totalMs` covers wall-clock while `stages` covers only profiled callbacks — the gap is undocumented | Add `unattributedMs` field or a JSDoc comment explaining `sum(stages) <= totalMs` |
| `session-start-profiler.ts:57-64` | `O_APPEND \| O_NOFOLLOW` may silently disable `O_NOFOLLOW` on Linux — the `lstatSync` check still provides TOCTOU protection, but the kernel-level guard is weaker than intended | Drop `O_APPEND`, use `lseek(fd, 0, SEEK_END)` + `write()` instead, or add post-open inode comparison |

_— qwen3.7-max via Qwen Code /review_

**Comment by wenshao** [comment]
@qwen-code /triage

**Comment by doudouOUC** [comment]
@wenshao addressed the suggestion summary across the follow-up commits.

Handled in tests:
- Added coverage for double `finish()` so only one record is written.
- Added coverage for throwing `timeSync()` preserving `failedStage` and rethrowing.
- Added assertions that each profiler from the startup/resume/clear source test is finalized exactly once.

Handled in latest commit `228028929`:
- Added a best-effort debug log when profiler record writing fails. The log is intentionally bounded to error name/code only, so it does not include paths, prompts, session ids, or profiling payload data.
- Renamed the stage from `deferred_reminder_resolution` to `deferred_reminder_setup`, since it covers both resolving the current deferred reminders and remembering the announced set.

I intentionally kept the profiler file write synchronous. This path is internal opt-in only (`QWEN_CODE_PROFILE_SESSION_START=1`) and writes one JSONL line; keeping it sync makes the measurement flush deterministic and avoids fire-and-forget lifecycle loss on short-lived CLI runs. The write remains best-effort: errors are swallowed and now leave a bounded debug breadcrumb when debug logging is active.

Verification run locally:
- `cd packages/core && npx vitest run src/core/session-start-profiler.test.ts src/core/client.test.ts`
- `npm run build`
- `npm run typecheck`
- `npm run lint`

File: packages/core/src/core/session-start-profiler.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+  const dir = path.join(Storage.getRuntimeBaseDir(), 'session-start-perf');
+  fs.mkdirSync(dir, { recursive: true, mode: 0o700 });
+  try {
+    fs.chmodSync(dir, 0o700);
+  } catch {
+    // Best-effort hardening on filesystems without POSIX chmod semantics.
+  }
+  const filename = `session-start-${record.timestamp.slice(0, 10)}.jsonl`;
+  const filePath = path.join(dir, filename);
+  const fd = fs.openSync(filePath, 'a', 0o600);
```
Comment (wenshao) [inline packages/core/src/core/session-start-profiler.ts]: **[Critical]** `fs.openSync(filePath, 'a', 0o600)` follows a pre-existing symlink at the deterministic daily profile path, and the subsequent `fchmodSync`/append operate on that target. If the runtime output directory is project-controlled or otherwise tampered with, enabling the profiler can append JSONL to an arbitrary user-writable file and chmod it to `0600`. Nearby durable write paths already use no-follow semantics for this exact class of issue; this writer should reject symlinked profiler directories/files before appending.

A fix should guard the profile directory and leaf file with no-follow/lstat checks before opening, for example:

```ts
const dirStat = fs.lstatSync(dir);
if (!dirStat.isDirectory() || dirStat.isSymbolicLink()) {
  throw new Error('session-start profiler path must be a real directory');
}

const noFollow =
  typeof fs.constants.O_NOFOLLOW === 'number' ? fs.constants.O_NOFOLLOW : 0;
const fd = fs.openSync(
  filePath,
  fs.constants.O_APPEND |
    fs.constants.O_CREAT |
    fs.constants.O_WRONLY |
    noFollow,
  0o600,
);
```

Also add a regression test that pre-places `session-start-YYYY-MM-DD.jsonl` as a symlink and proves the target is not modified.

_— gpt-5 via Qwen Code /review_

**Review by wenshao — CHANGES_REQUESTED** [review]

**Review by wenshao — COMMENTED** [review]
Reviewed — no blockers. Suggestion-level recommendations are in the **Suggestion summary** comment below.

**Comment by wenshao** [comment]
@qwen-code /triage

**Review by wenshao — DISMISSED** [review]

**Comment by wenshao** [comment]
@qwne-code /triage

File: packages/core/src/core/session-start-profiler.ts [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        ...(this.failedStage ? { failedStage: this.failedStage } : {}),
+      };
+
+      this.writeRecord(record);
+    } catch (error) {
+      const code = isNodeError(error) ? error.code : undefined;
+      try {
+        debugLogger.debug('session-start-profiler write failed', {
+          name: error instanceof Error ? error.name : typeof error,
+          message: error instanceof Error ? error.message : undefined,
```
[the remark this answers is not in the document]
Comment (doudouOUC) [reply packages/core/src/core/session-start-profiler.ts:227]: [codex] Not adopting this suggestion. The profiler safety boundary intentionally excludes file paths and working directories from emitted diagnostics; debug logs are also a diagnostic output channel. Keeping the write-failure breadcrumb bounded to error name/message/code preserves that boundary while still distinguishing common failures such as ENOSPC, EACCES, or symlink rejection.

**Comment by wenshao** [comment]
@qwne-code /triage

**Comment by wenshao** [comment]
## ✅ Maintainer verification — Linux end-to-end (covers the untested matrix row)

I built real tests for this PR in an isolated worktree off `03b00ef` with a full `npm ci`, then drove the **actual TUI under tmux** plus focused unit tests. The PR matrix marks **🐧 Linux ⚠️ not tested**; this run is entirely on **Linux** (Debian 13, kernel 6.12, Node v22.22.2), so it fills that gap — including the POSIX-only hardening (`O_NOFOLLOW`, `0700`/`0600`, `lstat` symlink guards) that never executes on the author's macOS-vs-Windows-only paths.

**Verdict: LGTM.** The env gate, bounded & non-sensitive output, best-effort write, symlink-attack defense, and restrictive permissions all hold on Linux, and a disabled session is unaffected (no output, no high-res clock read).

---

### 1) Focused unit tests — 255/255 on Linux
`cd packages/core && npx vitest run src/core/session-start-profiler.test.ts src/core/client.test.ts`
```
 ✓ src/core/session-start-profiler.test.ts (19 tests)
 ✓ src/core/client.test.ts             (236 tests)

 Test Files  2 passed (2)
      Tests  255 passed (255)
```
The two symlink defenses are gated `it.skip` **only on win32**, so on Linux they actually run — and pass:
```
 ✓ session-start-profiler > does not write through a symlinked JSONL file
 ✓ session-start-profiler > does not write through a symlinked profile directory
```

### 2) Real session under tmux — profiler ON writes one bounded record
Booted the real TUI (`node packages/cli/dist/index.js --yolo`) with `QWEN_CODE_PROFILE_SESSION_START=1` and an isolated `QWEN_RUNTIME_DIR` against a mock OpenAI endpoint (no live LLM). `GeminiClient.startChat()` (`source=startup`) fired at boot and produced exactly one record at `session-start-perf/session-start-2026-07-07.jsonl`:
```json
{"timestamp":"2026-07-07T04:21:13.676Z","source":"startup","ok":true,"totalMs":46.78,"stages":{"tool_registry_warm":38.04,"resume_deferred_tool_reveal":0.04,"deferred_reminder_setup":0.2,"initial_chat_history":5.85,"skill_reminder_seed":0.03,"agent_reminder_seed":1.21,"system_instruction":0.64,"gemini_chat_construct":0.08,"orphan_tool_use_repair":0.09,"session_start_hook":0.07,"set_tools":0.32},"extraHistoryLength":0,"historyLength":1,"snapshotEntryCount":9,"deferredReminderCount":46}
```
Stage breakdown (sorted, % of `totalMs`):
```
  tool_registry_warm           ██████████████████████████████████   38.04 (81%)
  initial_chat_history         █████·····························    5.85 (13%)
  agent_reminder_seed          █·································    1.21 (3%)
  system_instruction           █·································    0.64 (1%)
  set_tools                    ··································    0.32 (1%)
  deferred_reminder_setup      ··································     0.2 (0%)
  orphan_tool_use_repair       ··································    0.09 (0%)
  gemini_chat_construct        ··································    0.08 (0%)
  session_start_hook           ··································    0.07 (0%)
  resume_deferred_tool_reveal  ··································    0.04 (0%)
  skill_reminder_seed          ··································    0.03 (0%)
```
**Useful signal for #6312:** on a realistic project (46 deferred tools, 9 skills) `tool_registry_warm` dominates at **~81%**, *not* `system_instruction` (which the light fixture in #6312 surfaced). An independent headless run agreed (`35.88ms`, ~88%). So the profiler is doing its job — it shows the hotspot shifts with config, and on non-trivial setups the next optimization target is tool-registry warm-up, not system-instruction construction.

### 3) Env gate (A/B) — disabled = zero output
Same session, only difference is the env var:
```
ENABLED  (=1)    → session-start-perf/session-start-2026-07-07.jsonl  written
DISABLED (unset) → no session-start-perf directory created; session runs identically
```

### 4) Output is bounded & non-sensitive
The record contains only static field/stage names + numbers. Extracting every string **value** in the JSON tree yields just the timestamp and the static source enum:
```
string values in record = ["2026-07-07T04:21:13.676Z","startup"]
→ no prompt, no path, no sessionId, no tool names, no hook output
```

### 5) Restrictive permissions (POSIX)
```
700  session-start-perf/                       (dir)
600  session-start-perf/session-start-2026-07-07.jsonl   (file)
```

### 6) Symlink-attack defense at runtime (O_NOFOLLOW + lstat) — on the real write path
Planted a symlink at the exact target filename pointing at a secret file, then ran a real profiled session:
```
session-start-2026-07-07.jsonl  ->  VICTIM-secret.txt
result:  session started OK (exit 0)
         VICTIM-secret.txt UNCHANGED  ("ORIGINAL-SECRET-CONTENTS-DO-NOT-OVERWRITE")
         symlink not followed; write error swallowed (best-effort — startup never breaks)
```

### Environment
- Worktree off `03b00ef`, real `npm ci`; ran the built `packages/cli/dist/index.js`.
- Linux (Debian 13, kernel 6.12.63), Node **v22.22.2**, npm 10.9.7.
- Mock OpenAI endpoint; isolated `HOME` + `QWEN_RUNTIME_DIR` so nothing leaks into real config.

### Scope note
This verifies the runtime **behavior/safety** claims end-to-end on Linux. It does not re-review every line, and (per the PR's own scope) does not benchmark an end-to-end `/session` latency improvement — the profiler only *measures*.

<details>
<summary><b>中文版（合并参考）</b></summary>

## ✅ 维护者验证 —— Linux 端到端（补齐未测的矩阵行）

我在 `03b00ef` 上用独立 worktree + 完整 `npm ci` 为这个 PR 构建了真实测试，然后在 **tmux 下驱动真实 TUI**，并跑了聚焦单测。PR 矩阵里 **🐧 Linux ⚠️ not tested**，本次验证完全在 **Linux**（Debian 13，内核 6.12，Node v22.22.2）上完成，正好补上这一行 —— 包括只在 POSIX 上执行的加固逻辑（`O_NOFOLLOW`、`0700`/`0600`、`lstat` 符号链接校验），这些在作者只测的 macOS 路径之外。

**结论：LGTM。** 环境变量开关、有边界且非敏感的输出、best-effort 写入、符号链接攻击防御、受限权限在 Linux 上全部成立；关闭时会话行为完全不受影响（不写文件、不读高精度时钟）。

**1) 聚焦单测 —— Linux 上 255/255**
`session-start-profiler.test.ts (19)` + `client.test.ts (236)`，全部通过。其中两个符号链接防御用例只在 win32 上 `it.skip`，因此在 Linux 上是真实执行并通过的：
- `does not write through a symlinked JSONL file`
- `does not write through a symlinked profile directory`

**2) tmux 下真实会话 —— 开启 profiler 会写出一条有边界的记录**
用 `node packages/cli/dist/index.js --yolo` 启动真实 TUI，设置 `QWEN_CODE_PROFILE_SESSION_START=1` + 隔离的 `QWEN_RUNTIME_DIR`，对接 mock OpenAI（无真实 LLM）。启动时 `startChat()`（`source=startup`）触发，恰好写出一条记录（内容见英文版 JSON）。
阶段耗时排序显示：在真实工程（46 个 deferred tools、9 个 skills）下 **`tool_registry_warm` 占约 81%**，而不是 `system_instruction`（#6312 的轻量 fixture 里主导的那个）；一次独立 headless 运行也一致（`35.88ms`，约 88%）。这正说明 profiler 起到了作用：热点会随配置变化，在非平凡配置下下一步优化目标是 tool registry 预热，而非 system instruction 构造 —— 对 #6312 是有用的实测证据。

**3) 开关 A/B —— 关闭即零输出**
仅差一个环境变量：开启 → 写出 JSONL；关闭（不设置）→ 不创建 `session-start-perf` 目录，会话行为完全相同。

**4) 输出有边界且不含敏感信息**
整条记录里所有字符串**值**只有两个：时间戳和静态的 source 枚举 `["2026-07-07T04:21:13.676Z","startup"]` —— 没有 prompt、路径、sessionId、工具名、hook 输出。

**5) 受限权限（POSIX）**
目录 `700`、文件 `600`，与设计一致。

**6) 运行期符号链接攻击防御（O_NOFOLLOW + lstat）—— 在真实写入路径上**
在目标文件名处放置一个指向机密文件的符号链接，再跑一次开启 profiler 的真实会话：会话正常启动（exit 0），机密文件**未被改写**，符号链接未被跟随，写入错误被吞掉（best-effort，绝不影响启动）。

**环境**：`03b00ef` worktree + 真实 `npm ci`；Linux（Debian 13，内核 6.12.63），Node v22.22.2，npm 10.9.7；mock OpenAI，隔离 `HOME`/`QWEN_RUNTIME_DIR`。

**范围说明**：本次在 Linux 上端到端验证的是运行期**行为/安全**；未逐行复审，也（按 PR 自身范围）未对 `/session` 端到端延迟做基准 —— profiler 只负责测量。

</details>

**Review by wenshao — APPROVED** [review]

**Comment by wenshao** [comment]
@qwen-code /triage
